ophyd/_caproto_shim.py

```diff
@@ -1,11 +1,120 @@
+                         auto_monitor=auto_monitor, count=count,
+                         connection_timeout=connection_timeout,
+                         connection_callback=connection_callback,
+                         callback=callback, access_callback=access_callback,
+                         context=context)
+
+    def add_callback(self, callback=None, index=None, run_now=False,
+                     with_ctrlvars=True, **kw):
+        callback = wrap_callback(_dispatcher, 'monitor', callback)
+        return super().add_callback(callback=callback, index=index,
+                                    run_now=run_now,
+                                    with_ctrlvars=with_ctrlvars, **kw)
+
+    def put(self, value, wait=False, timeout=30.0, use_complete=False,
+            callback=None, callback_data=None):
+        callback = wrap_callback(_dispatcher, 'get_put', callback)
+        return super().put(value, wait=wait, timeout=timeout,
+                           use_complete=use_complete, callback=callback,
+                           callback_data=callback_data)
+
+
+def get_pv(pvname, form='time', connect=False, context=None, timeout=5.0,
+           connection_callback=None, access_callback=None, callback=None,
+           **kwargs):
+    """Get a PV from PV cache or create one if needed.
+
+    Parameters
+    ---------
+    form : str, optional
+        PV form: one of 'native' (default), 'time', 'ctrl'
+    connect : bool, optional
+        whether to wait for connection (default False)
+    context : int, optional
+        PV threading context (defaults to current context)
+    timeout : float, optional
+        connection timeout, in seconds (default 5.0)
+    """
+    if context is None:
+        context = PV._default_context
+
+    pv = PV(pvname, form=form, connection_callback=connection_callback,
+            access_callback=access_callback, callback=callback,
+            **kwargs)
+    if connect:
+        pv.wait_for_connection(timeout=timeout)
+    return pv
+
+
+def setup(logger):
+    '''Setup ophyd for use
+
+    Must be called once per session using ophyd
+    '''
+    # It's important to use the same context in the callback dispatcher
+    # as the main thread, otherwise not-so-savvy users will be very
+    # confused
+    global _dispatcher
+
+    if _dispatcher is not None:
+        logger.debug('ophyd already setup')
+        return
+
+    pyepics_compat._get_pv = pyepics_compat.get_pv
+    pyepics_compat.get_pv = get_pv
+
+    def _cleanup():
+        '''Clean up the ophyd session'''
+        global _dispatcher
+        if _dispatcher is None:
+            return
+
+        pyepics_compat.get_pv = pyepics_compat._get_pv
+
+        logger.debug('Performing ophyd cleanup')
+        if _dispatcher.is_alive():
+            logger.debug('Joining the dispatcher thread')
+            _dispatcher.stop()
+
+        _dispatcher = None
+
+    logger.setLevel('DEBUG')
```

Comment:
I don't think we should set logger levels. Feel free to punt this to some future comprehensive review of loggers in ophyd, if you want. Now that I finally understood them for caproto, I want to clean up ophyd and bluesky logging....

Reply:
💯% agreed and good catch - this was for my debugging purposes and needs to be removed.